packages/dds/tree/src/simple-tree/discrepancies.ts

```diff
@@ -158,56 +198,71 @@ function getNodeSchemaType(nodeSchema: TreeNodeStoredSchema): SchemaFactoryNodeK
-	yield* getFieldDiscrepancies(
-		view.rootFieldSchema,
-		stored.rootFieldSchema,
-		undefined,
-		undefined,
-	);
+	yield* getFieldDiscrepancies(view, stored.rootFieldSchema, undefined, undefined);
+
+	const storedAllowedTypes = stored.nodeSchema;
+
+	// collect all annotated allowed types from the view schema
+	const annotatedAllowedTypes: AnnotatedAllowedType<TreeNodeSchema>[] = [];
+	walkFieldSchema(view, {
+		allowedTypes: (allowedTypes) => {
+			annotatedAllowedTypes.push(...allowedTypes.types);
+		},
+	});
+
+	const viewAllowedTypes = new Map<TreeNodeSchemaIdentifier, TreeNodeSchema>();
+	for (const annotatedAllowedType of annotatedAllowedTypes) {
+		const { type } = annotatedAllowedType;
+		// map view schema identifiers to the field schemas to make access in the stored schema pass more efficient
+		const identifier: TreeNodeSchemaIdentifier = brand(type.identifier);
+		viewAllowedTypes.set(identifier, type);
+
+		const storedSchema = stored.nodeSchema.get(identifier);
+
+		// if the view schema has an allowed type that's not in the stored schema
+		if (!storedAllowedTypes.has(identifier) || storedSchema === undefined) {
+			const viewType = getViewNodeSchemaType(type);
+			// TODO does it make sense to have this mismatch when there will also be an allowedTypes mismatch?
+			yield {
+				identifier,
+				mismatch: "nodeKind",
+				view: viewType,
+				stored: undefined,
+			};
+		} else {
+			yield* getNodeDiscrepancies(identifier, annotatedAllowedType, storedSchema);
+		}
+	}
 
-	for (const result of compareMaps(view.nodeSchema, stored.nodeSchema)) {
-		switch (result.type) {
-			case "aExtra": {
-				const viewType = getNodeSchemaType(result.value);
-				yield {
-					identifier: result.key,
-					mismatch: "nodeKind",
-					view: viewType,
-					stored: undefined,
-				};
-				break;
-			}
-			case "bExtra": {
-				const storedType = getNodeSchemaType(result.value);
-				yield {
-					identifier: result.key,
-					mismatch: "nodeKind",
-					view: undefined,
-					stored: storedType,
-				};
-				break;
-			}
-			case "both": {
-				yield* getNodeDiscrepancies(result.key, result.valueA, result.valueB);
-				break;
-			}
-			default:
-				break;
+	for (const identifier of storedAllowedTypes.keys()) {
+		if (!viewAllowedTypes.has(identifier)) {
+			const storedType = getStoredNodeSchemaType(
+				stored.nodeSchema.get(identifier) ??
+					fail(
+						"Stored schema should have a schema for an identifier present in the root schema types",
+					),
+			);
+			yield {
+				identifier,
+				mismatch: "nodeKind",
+				view: undefined,
+				stored: storedType,
+			};
 		}
 	}
 }
 
 function* getNodeDiscrepancies(
 	identifier: TreeNodeSchemaIdentifier,
-	view: TreeNodeStoredSchema,
+	{ type: view }: AnnotatedAllowedType<TreeNodeSchema>,
 	stored: TreeNodeStoredSchema,
 ): Iterable<Discrepancy> {
-	const viewType = getNodeSchemaType(view);
-	const storedType = getNodeSchemaType(stored);
+	const viewType = getViewNodeSchemaType(view);
```

Comment:
Rather than using getViewNodeSchemaType, I think just switching over the node kinds, using the narrowing functions (like isArrayNodeSchema) in an else/if chain, or both (like in unhydratedFlexTreeFromInsertableNode) would be a better pattern here.

If you narrow the schema to their more specific types, you can access things like `MapNodeSchema`'s `.info` field or  ObjectNodeSchemaPrivate's storedKeyToPropertyKey and allowUnknownOptionalFields directly in a type safe way.

Reply:
I think I addressed this? Lmk if this is what you had in mind.